Automattic/wp-calypso
# CPT / State: Display trash notice after save request completes successfully

This pull request seeks to explore one option for displaying a notice in response to a successful network request to save a post. It also seeks to refactor notices state to better accommodate this use-case and to adhere to guidelines adopted since its original implementation (e.g. action type verb suffixing).

![trash](https://cloud.githubusercontent.com/assets/1779930/16347105/ee1d3674-3a18-11e6-957e-078aa8d5ecb7.gif)

__Implementation notes:__

I considered a few options for implementing notices:

- Dispatch `successNotice` in the promise resolution of `savePost`
  - Pros:
    - Colocation to related actions leading to notice creation
    - Notices shown even if navigating away from screen
  - Cons:
    - Difficult to differentiate user interaction from background process where notice may be undesirable (maybe options argument?)
    - Doesn't seem possible to track notice state like count (i.e. "2 posts trashed")
- Track success and failure in a separate reducer specific to posts, which components monitor and dispatch notice when `componentWillReceiveProps` reveals that a success or failure has occurred
  - Pros:
    - Feels most pure in terms of one-way-data-flow
    - If "notices shown after navigating away" is argued to be a bad thing, then this is most obvious path to achieving the opposite effect
  - Cons:
    - Notice won't be shown if user navigates away (or component is otherwise unmounted)
    - Seems difficult to implement correctly, and likely more code to do so than other solutions
- [As implemented here, ] Monitor actions from the notices reducer
  - Pros: 
    - If considering action types as global, appropriate for notices reducer to "watch" for actions which should cause a notice to be shown
    - All behaviors related to notices mostly within a single file
    - Notices shown even if navigating away from screen
  - Cons:
    - Splits action creator invoking the save from notices behavior (maybe a pro?)
    - Should notices reducer concern itself with actions from other parts of state? (seems an argument against actions as being global)

__Testing instructions:__

Ensure Mocha tests pass:

```
npm run test-client
```

Verify that no regressions occurred in refactorings to expected behavior for [global notices](http://calypso.localhost:3000/devdocs/design/global-notices).

Verify that on the [custom post types listing screen](http://calypso.localhost:3000/types/post), when trashing a post, a notice is shown if it completes successfully. Verify also that if trashing a second post while a notice is currently visible, the text will change to reflect "X posts" trashed.

/cc @artpi @gwwar @timmyc 

Test live: https://calypso.live/?branch=add/cpt-trash-notice
